kratos/utilities/sparse_matrix_multiplication_utility.h

```diff
@@ -683,6 +683,267 @@ class SparseMatrixMultiplicationUtility
+                                const SizeType size_system_1 = r_matrix.size1();
+                                const SizeType size_system_2 = r_matrix.size2();
+                                TMatrix transpose(size_system_2, size_system_1);
+                                TransposeMatrix<TMatrix, TMatrix>(transpose, r_matrix);
+                                ComputeNonZeroBlocks<TMatrix>(transpose, k, matrix_cols_aux);
+                            #ifdef KRATOS_DEBUG
+                                ComputeNonZeroBlocks<TMatrix>(transpose, k, partial_matrix_cols_aux);
+                            #endif
+                            } else {
+                                ComputeNonZeroBlocks<TMatrix>(r_matrix, k, matrix_cols_aux);
+                            #ifdef KRATOS_DEBUG
+                                ComputeNonZeroBlocks<TMatrix>(r_matrix, k, partial_matrix_cols_aux);
+                            #endif
+                            }
+                        }
+                    #ifdef KRATOS_DEBUG
+                        check_non_zero_blocks(i, j) += partial_matrix_cols_aux;
+                    #endif
+                    }
+                    IndexType& r_matrix_ptr_value = matrix_ptr[std::accumulate(row_sizes.begin(), row_sizes.begin() + i, 0) + k + 1];
+                    #pragma omp atomic
+                    r_matrix_ptr_value += matrix_cols_aux;
+
+                #ifdef KRATOS_DEBUG
+                    #pragma omp atomic
+                    check_non_zero += matrix_cols_aux;
+                #endif
+                }
+            }
+        }
+
+        // Auxiliar values
+        std::partial_sum(matrix_ptr, matrix_ptr + nrows + 1, matrix_ptr);
+        const SizeType nonzero_values = matrix_ptr[nrows];
+
+    #ifdef KRATOS_DEBUG
+        SizeType total_nnz = 0;
+        for (int i=0; i<static_cast<int>(number_of_rows_blocks); ++i) {
+            for (int j=0; j<static_cast<int>(number_of_columns_blocks); ++j) {
+                const SizeType block_nnz = rMatricespBlocks(i, j)->nnz();
+                KRATOS_ERROR_IF_NOT(check_non_zero_blocks(i, j) == block_nnz) << "Inconsistent number of non-zero values. Check 0: " << block_nnz << " vs " << check_non_zero_blocks(i, j) << ". Block: " << i << ", " << j << std::endl;
+                total_nnz += block_nnz;
+            }
+        }
+        KRATOS_ERROR_IF_NOT(check_non_zero == total_nnz) << "Inconsistent number of non-zero values. Check 1: " << total_nnz << " vs " << check_non_zero << std::endl;
+        KRATOS_ERROR_IF_NOT(nonzero_values == total_nnz) << "Inconsistent number of non-zero values. Check 2: " << total_nnz << " vs " << nonzero_values << std::endl;
+    #endif
+
+        // Initialize matrix with the corresponding non-zero values
+        rMatrix = TMatrix(nrows, ncols, nonzero_values);
+
+        // Fill the new matrix
+        double* Matrix_values = rMatrix.value_data().begin();
```

Comment:
here you are assuming it is a compressed_matrix...so the use of template for matrix type is effectively not useful

Reply:
This is done in the rest of sparse_matrix_multiplication_utility.h. BTW, the builders and solvers assume also that it is a compressed matrix and we still consider templates for the spaces...

Reply:
as far as i know this is not the case. solvers are used both for ublas and for trilinos...they would not work otherwise